pkg/volume/metadata/metadata.go

```diff
@@ -0,0 +1,271 @@
+	return plugin.newBuilderInternal(spec, pod, opts, mounter)
+}
+
+func (plugin *metadataPlugin) newBuilderInternal(spec *volume.Spec, pod *api.Pod, opts volume.VolumeOptions, mounter mount.Interface) (volume.Builder, error) {
+	v := &metadataVolume{volName: spec.Name,
+		pod:     pod,
+		plugin:  plugin,
+		opts:    &opts,
+		mounter: mounter}
+	v.fieldReferenceFileNames = make(map[string]string)
+	for _, fileInfo := range spec.VolumeSource.Metadata.Items {
+		v.fieldReferenceFileNames[fileInfo.FieldRef.FieldPath] = fileInfo.Name
+	}
+	return v, nil
+}
+
+func (plugin *metadataPlugin) NewCleaner(volName string, podUID types.UID, mounter mount.Interface) (volume.Cleaner, error) {
+	return plugin.newCleanerInternal(volName, podUID, mounter)
+}
+
+func (plugin *metadataPlugin) newCleanerInternal(volName string, podUID types.UID, mounter mount.Interface) (volume.Cleaner, error) {
+	return &metadataVolume{volName: volName,
+		pod:     &api.Pod{ObjectMeta: api.ObjectMeta{UID: podUID}},
+		plugin:  plugin,
+		mounter: mounter}, nil
+}
+
+// metadataVolume handles retrieving metadata from the API server
+// and placing them into the volume on the host.
+type metadataVolume struct {
+	volName                 string
+	fieldReferenceFileNames map[string]string
+	pod                     *api.Pod
+	plugin                  *metadataPlugin
+	opts                    *volume.VolumeOptions
+	mounter                 mount.Interface
+}
+
+// SetUp puts in place the volume plugin.
+// This function is not idempotent by design. We want the data to be refreshed periodically.
+// The internal sync interval of kubelet will drive the refresh of data.
+func (m *metadataVolume) SetUp() error {
+	return m.SetUpAt(m.GetPath())
+}
+
+// This is the spec for the volume that this plugin wraps.
+var wrappedVolumeSpec = &volume.Spec{
+	Name:         "not-used",
+	VolumeSource: api.VolumeSource{EmptyDir: &api.EmptyDirVolumeSource{Medium: api.StorageMediumMemory}},
+}
+
+// readFile reads full path file name and return it as a string
+func readFile(fileName string) string {
+	if data, err := ioutil.ReadFile(fileName); err == nil {
+		return string(data)
+	}
+	return ""
+}
+
+// sortLines sorts the strings generated from map based metadata
+// (annotations and labels)
+func sortLines(values string) string {
+	splitted := strings.Split(values, "\n")
+	sort.Strings(splitted)
+	return strings.Join(splitted, "\n")
+}
+
+// isFileToGenerate compares actual file with the new values. If
+// different (or the file does not exist) return true
+func isFileToGenerate(fileName, values string) bool {
+	if _, err := os.Lstat(fileName); os.IsNotExist(err) {
+		return true
+	}
+	return readFile(fileName) != values
+}
+
+const currentDir = ".current"
+
+// writeDataInTmpDir writes metadata in file
+func (m *metadataVolume) writeDataInTmpDir() (string, error) {
+	errlist := []error{}
+	tempDir, err := ioutil.TempDir(m.GetPath(), "")
```

Comment:
@thockin @sdminonne 

Tim's right -- the directory names `ioutil.TempDir` creates result in files that are visible in the directory.  Instead, can we create a directory named `.<timestamp>` instead?